<a href="https://colab.research.google.com/github/Shivani-ks-14/Full-Stack-Hotel-Management-Application/blob/main/Sigma_Lats_Task_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [49]:
# importing libraries
import pandas as pd

In [52]:
# importing dataset
from google.colab import files
uploaded = files.upload()

Saving edip_dataset.csv to edip_dataset (1).csv


In [54]:
df = pd.read_csv('edip_dataset.csv')

In [55]:
df.head()

,country_name,iso3_code,year,income_group,region,gdp_growth_pct,gdp_per_capita_usd,inflation_cpi_pct,unemployment_rate,govt_debt_pct_gdp,...,gini_index,poverty_headcount_190,fdi_net_inflows_pct_gdp,trade_openness_pct_gdp,internet_access_pct,govt_expenditure_pct_gdp,urban_population_pct,hdi_score,hdi_tier,is_projection
0,Afghanistan,AFG,1995,Low,"Middle East, North Africa, Afghanistan & Pakistan",NaN,NaN,NaN,7.868,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,17.834385,NaN,NaN,0
1,Afghanistan,AFG,1996,Low,"Middle East, North Africa, Afghanistan & Pakistan",NaN,NaN,NaN,7.878,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,17.959649,NaN,NaN,0
2,Afghanistan,AFG,1997,Low,"Middle East, North Africa, Afghanistan & Pakistan",NaN,NaN,NaN,7.863,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,18.092857,NaN,NaN,0
3,Afghanistan,AFG,1998,Low,"Middle East, North Africa, Afghanistan & Pakistan",NaN,NaN,NaN,7.869,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,18.235814,NaN,NaN,0
4,Afghanistan,AFG,1999,Low,"Middle East, North Africa, Afghanistan & Pakistan",NaN,NaN,NaN,7.877,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,18.390327,NaN,NaN,0


**EXPLORATORY DATA ANALYTICS**

This dataset provides country-wise economic and development indicators across different years.
The objective is to predict the HDI (Human Development Index).
Predicting the HDI score is a regression problem and predicting the HDI tier is a classification problem.

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6020 entries, 0 to 6019
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   country_name              6020 non-null   object 
 1   iso3_code                 6020 non-null   object 
 2   year                      6020 non-null   int64  
 3   income_group              6020 non-null   object 
 4   region                    6020 non-null   object 
 5   gdp_growth_pct            5696 non-null   float64
 6   gdp_per_capita_usd        5785 non-null   float64
 7   inflation_cpi_pct         4922 non-null   float64
 8   unemployment_rate         5179 non-null   float64
 9   govt_debt_pct_gdp         1291 non-null   float64
 10  current_account_pct_gdp   4751 non-null   float64
 11  fiscal_balance_pct_gdp    3127 non-null   float64
 12  gini_index                1986 non-null   float64
 13  poverty_headcount_190     1986 non-null   float64
 14  fdi_net_

The given dataset has 22 features (columns) and 6020 records (rows).
All columns seem to have appropriate datatypes.

Here we will drop the country_name column as we already have country code, so the features are redundant

In [57]:
# dropping country name feature
df.drop(columns = ['country_name'],inplace = True)

In [58]:
df.describe()

,year,gdp_growth_pct,gdp_per_capita_usd,inflation_cpi_pct,unemployment_rate,govt_debt_pct_gdp,current_account_pct_gdp,fiscal_balance_pct_gdp,gini_index,poverty_headcount_190,fdi_net_inflows_pct_gdp,trade_openness_pct_gdp,internet_access_pct,govt_expenditure_pct_gdp,urban_population_pct,hdi_score,hdi_tier,is_projection
count,6020.000000,5696.000000,5785.000000,4922.000000,5179.000000,1291.000000,4751.000000,3127.000000,1986.000000,1986.000000,5283.000000,4847.000000,5344.000000,3197.000000,6020.000000,5438.000000,5438.000000,6020.000000
mean,2008.500000,3.501155,14556.409122,9.746885,8.228729,55.055740,-2.427599,-1.452086,37.141692,10.479104,8.316828,88.177842,32.004105,27.042167,58.422694,0.668775,0.476278,0.071429
std,8.078418,6.395307,23194.999525,76.795669,6.099334,44.258772,14.020584,9.289009,8.434888,18.173242,58.161037,56.864049,31.491168,20.978966,24.143737,0.189242,0.499483,0.257561
min,1995.000000,-54.402093,62.143814,-16.859691,0.100000,-1.170726,-147.997312,-52.849600,23.000000,0.000000,-1303.108267,0.020999,0.000000,2.045768,7.168294,0.110440,0.000000,0.000000
25%,2001.750000,1.307714,1315.736346,1.640548,3.806500,30.036443,-7.247260,-4.104529,30.800000,0.300000,1.023446,53.730006,2.902000,17.355240,38.075708,0.537428,0.000000,0.000000
50%,2008.500000,3.649460,4536.733121,3.601412,6.491000,49.157455,-2.839518,-1.931081,35.200000,1.600000,2.720999,76.398690,20.805000,25.527839,59.277145,0.686153,0.000000,0.000000
75%,2015.250000,5.925031,19009.077441,7.537890,11.137000,69.032329,1.717798,0.157273,42.075000,11.400000,5.639109,105.732129,59.929301,34.515740,78.030878,0.808559,1.000000,0.000000
max,2022.000000,149.972963,226052.001905,4145.105982,38.800000,865.873243,311.746071,238.548732,65.800000,94.900000,1709.827224,863.195099,100.000000,758.958242,100.000000,0.994352,1.000000,1.000000


Handling nulls and duplicates

In [59]:
# check whether duplicate rows exist
any(df.duplicated())

False

In [60]:
# check the no of null values in each column
df.isnull().sum(axis=0)

,0
iso3_code,0
year,0
income_group,0
region,0
gdp_growth_pct,324
gdp_per_capita_usd,235
inflation_cpi_pct,1098
unemployment_rate,841
govt_debt_pct_gdp,4729
current_account_pct_gdp,1269


In [61]:
# percentage of missing values in each column
missing_pct = (df.isnull().sum(axis=0)/len(df))*100
print(missing_pct)

iso3_code                    0.000000
year                         0.000000
income_group                 0.000000
region                       0.000000
gdp_growth_pct               5.382060
gdp_per_capita_usd           3.903654
inflation_cpi_pct           18.239203
unemployment_rate           13.970100
govt_debt_pct_gdp           78.554817
current_account_pct_gdp     21.079734
fiscal_balance_pct_gdp      48.056478
gini_index                  67.009967
poverty_headcount_190       67.009967
fdi_net_inflows_pct_gdp     12.242525
trade_openness_pct_gdp      19.485050
internet_access_pct         11.229236
govt_expenditure_pct_gdp    46.893688
urban_population_pct         0.000000
hdi_score                    9.667774
hdi_tier                     9.667774
is_projection                0.000000
dtype: float64


We observe that govt_debt_pct_gdp, gini_index, and poverty_headcount_190 contain a large percentage of missing values (>50%).
Here gini_index and poverty_headcount_190 are useful indicators in predicting HDI, poverty or development. We may use certain intelligent imputation methods like country wise imputation or interpolation over years to fill in the null values. But we end up with a huge amount of artificial data.  
So we shall drop these features and evaluate our model performance.


In [62]:
# drop columns with more than 50% missing values
df = df.loc[:,df.isnull().mean()<0.5]

In [63]:
# recheck percentage of missing values
missing_pct = (df.isnull().sum(axis=0)/len(df))*100
print(missing_pct)

iso3_code                    0.000000
year                         0.000000
income_group                 0.000000
region                       0.000000
gdp_growth_pct               5.382060
gdp_per_capita_usd           3.903654
inflation_cpi_pct           18.239203
unemployment_rate           13.970100
current_account_pct_gdp     21.079734
fiscal_balance_pct_gdp      48.056478
fdi_net_inflows_pct_gdp     12.242525
trade_openness_pct_gdp      19.485050
internet_access_pct         11.229236
govt_expenditure_pct_gdp    46.893688
urban_population_pct         0.000000
hdi_score                    9.667774
hdi_tier                     9.667774
is_projection                0.000000
dtype: float64


Now the columns with very high missing values are removed. The dataset in much more manageable.
Now we shall handle the remaining missing values using imputation.

We will do the imputation in stages


*   Interpolation by country - to preserve yearly trends
*   country wise median - to preserve country characteristics
*   Region wise median - use region similarity
*   global median fall back - to remove remaining null values

In [64]:
# sort the data properly
df = df.sort_values(['iso3_code','year'])

In [65]:
# separate numerical and categorical columns
num_cols = df.select_dtypes(include = ['float64','int64']).columns
cat_cols = df.select_dtypes(include = 'object').columns

In [66]:
num_cols

Index(['year', 'gdp_growth_pct', 'gdp_per_capita_usd', 'inflation_cpi_pct',
       'unemployment_rate', 'current_account_pct_gdp',
       'fiscal_balance_pct_gdp', 'fdi_net_inflows_pct_gdp',
       'trade_openness_pct_gdp', 'internet_access_pct',
       'govt_expenditure_pct_gdp', 'urban_population_pct', 'hdi_score',
       'hdi_tier', 'is_projection'],
      dtype='object')

In [67]:
cat_cols

Index(['iso3_code', 'income_group', 'region'], dtype='object')

We separate numerical and categorical columns because we impute numerical columns with median, and categorical columns with mode.
Here the categorical columns have no null values, so no action required.

In [68]:
# interpolation by country
for col in num_cols:
  df[col] = df.groupby('iso3_code')[col]\
  .transform(lambda x: x.interpolate())

In [69]:
# country wise median imputation
for col in num_cols:
  df[col] = df.groupby('iso3_code')[col]\
  .transform(lambda x: x.fillna(x.median()))

In [70]:
# region wise median imputation
for col in num_cols:
  df[col] = df.groupby('region')[col]\
  .transform(lambda x: x.fillna(x.median()))

In [71]:
# global median fall back
for col in num_cols:
  df[col] = df[col].fillna(df[col].median())

In [72]:
# verify that there's no more null values
df.isnull().sum(axis=0)

,0
iso3_code,0
year,0
income_group,0
region,0
gdp_growth_pct,0
gdp_per_capita_usd,0
inflation_cpi_pct,0
unemployment_rate,0
current_account_pct_gdp,0
fiscal_balance_pct_gdp,0


Therefore we have handled all the null values

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6020 entries, 252 to 6019
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   iso3_code                 6020 non-null   object 
 1   year                      6020 non-null   int64  
 2   income_group              6020 non-null   object 
 3   region                    6020 non-null   object 
 4   gdp_growth_pct            6020 non-null   float64
 5   gdp_per_capita_usd        6020 non-null   float64
 6   inflation_cpi_pct         6020 non-null   float64
 7   unemployment_rate         6020 non-null   float64
 8   current_account_pct_gdp   6020 non-null   float64
 9   fiscal_balance_pct_gdp    6020 non-null   float64
 10  fdi_net_inflows_pct_gdp   6020 non-null   float64
 11  trade_openness_pct_gdp    6020 non-null   float64
 12  internet_access_pct       6020 non-null   float64
 13  govt_expenditure_pct_gdp  6020 non-null   float64
 14  urban_popul

Our dataset now has 18 features (columns) and 6020 records (rows)

In [74]:
# statistical summary of the numerical features
df.describe()

,year,gdp_growth_pct,gdp_per_capita_usd,inflation_cpi_pct,unemployment_rate,current_account_pct_gdp,fiscal_balance_pct_gdp,fdi_net_inflows_pct_gdp,trade_openness_pct_gdp,internet_access_pct,govt_expenditure_pct_gdp,urban_population_pct,hdi_score,hdi_tier,is_projection
count,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000,6020.000000
mean,2008.500000,3.416036,14658.213369,8.831574,7.985922,-2.830701,-1.513369,8.066673,89.432235,32.602088,26.061002,58.422694,0.674761,0.491860,0.071429
std,8.078418,6.298677,23050.147684,69.547022,5.725457,16.509810,7.105065,68.000110,55.392069,31.187106,23.257353,24.143737,0.187116,0.499975,0.257561
min,1995.000000,-54.402093,62.143814,-16.859691,0.100000,-147.997312,-52.849600,-1303.108267,0.020999,0.000000,2.045768,7.168294,0.110440,0.000000,0.000000
25%,2001.750000,1.287872,1335.550874,1.813389,3.779750,-6.921812,-3.079361,1.176682,57.325946,3.519689,16.678383,38.075708,0.550511,0.000000,0.000000
50%,2008.500000,3.484241,4698.398105,3.364561,6.881500,-2.658717,-2.216677,2.849656,76.479332,22.110300,22.584367,59.277145,0.695340,0.000000,0.000000
75%,2015.250000,5.818331,19217.319571,6.805940,10.267750,1.464080,0.061105,5.341118,103.605196,59.592550,32.141024,78.030878,0.815323,1.000000,0.000000
max,2022.000000,149.972963,226052.001905,4145.105982,38.800000,311.746071,238.548732,1709.827224,863.195099,100.000000,758.958242,100.000000,0.994352,1.000000,1.000000
